library/src/main/scala/scala/scalajs/macroImpls/UseAsMacros.scala

```diff
@@ -0,0 +1,304 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package scala.scalajs.macroImpls
+
+import scala.annotation.tailrec
+import scala.scalajs.js
+
+import Compat210._
+
+/** Macros for js.use(x).as[T]
+ *
+ *  This implements a structural typechecker conforming to the JavaScript calling
+ *  convention in Scala.js' export and facade system.
+ *
+ *  @author Tobias Schlatter
+ */
+object UseAsMacros {
```

Comment:
I don't think you are allowed to make macro implementations non-public. But I'll try.